kubelet/kuberuntime: switch to runc/libct

kolyshkin opened the pull request:
Related to discussion in #128157 - Let's ensure we use ONLY one cgroups library.

runc/libcontainer/cgroups.ParseCgroupFile is a universal function; for cgroup v2 it returns path to unified in a map with "" as a key.

Let's use it here, dropping cgroups dependency entirely.

Amends commit e86d02b60c1 / PR #128473.

/kind cleanup

```release-note
none
```

kolyshkin committed:
kubelet/kuberuntime: switch to runc/libct

runc/libcontainer/cgroups.ParseCgroupFile is a universal function; for
cgroup v2 it returns path to unified in a map with "" as a key.

Let's use it here, dropping cgroups dependency entirely.

Amends commit e86d02b60c1.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

dims commented:
/test all

kolyshkin commented:
I don't think the test failure is related.

/retest

bart0sh commented:
/triage accepted

bart0sh commented:
/cc @SergeyKanzhelev @kannon92

kannon92 commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -352,12 +351,12 @@ var swapControllerAvailable = func() bool {
 				return
 			}
-			p = filepath.Join("/sys/fs/cgroup", unified, "memory.swap.max")
+			p = filepath.Join("/sys/fs/cgroup", cm[""], "memory.swap.max")
```
This seems a bit esoteric..

Can you explain why we are grabbing cm[""] here?

SergeyKanzhelev replied:
(explain by adding a comment into code)

kolyshkin replied:
The answer is a click away (in my $EDITOR, at least) -- take a look at the [ParseCgroupFile docs](https://pkg.go.dev/github.com/opencontainers/runc/libcontainer/cgroups#ParseCgroupFile), it says:

> Note that for cgroup v2 unified hierarchy, there are no per-controller cgroup paths, so the resulting map will have a single element where the key is empty string ("") and the value is the cgroup path the <pid> is in.

Nevertheless, I've added the comment.

kolyshkin committed:
kubelet/kuberuntime: use sync.OnceValue

This was added to Go 1.21, and makes the code simpler.

(Best reviewed ignoring changes in amount of whitespace).

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
Added a separate second commit which makes use of `sync.OnceValue`.

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

SergeyKanzhelev commented:
/assign @dims

dims commented:
/approve
/lgtm

thanks @kolyshkin

pacoxu commented:
/milestone v1.32
approved before code freeze
